### ISSUE TITLE
MailParser block forever with email multipart / rfc822 with header only

### ISSUE THREAD

**Issue by peernohell** [nodemailer/mailparser#190]
Hi,

We have strange issue where mailparser never give back the mail content when the email is a multi part and one part is an rfc822 with only header.

I have write a test you can add in mail-parser-test.js:
```
    'Simple with part as rfc822 of headers only': test => {
        let encodedText =
            `Content-Type: multipart/mixed; boundary="ABC"

--ABC
Content-Type: message/rfc822

Content-Type: text/plain; charset=utf-8
Subject: OK
--ABC--`,
            mail = Buffer.from(encodedText, 'utf-8');

        let mailparser = new MailParser();
        mailparser.end(mail);
        mailparser.on('data', () => false);
        mailparser.on('end', () => {
            test.equal(mailparser.text, '\nSubject: OK\n\nok');
            test.done();
        });
    },
```

If I add content in the email every thing seems ok. 
Could you help me understand why mailparser block?

Thanks a lot 🙇‍♂️

**Comment by andris9** [comment]
Seems to be a bug in the parser, content is processed but callback is never triggered. Not sure when I have time to look at it

**Comment by peernohell** [comment]
I have investigate a bit.
What I have found is in mail-parser.js line 143:
```
        if (this.curnode && this.curnode.decoder && this.curnode.decoder.readable) {
            (this.curnode.contentStream || this.curnode.decoder).once('end', () => {
                finish();
            });
            this.curnode.decoder.end();
```
What is different from an email with content is that this.curnode.contentStream didn't exists. so we add the event on the decoder, but when we call decoder.end() as the decoder have `state.ending` and `state.finished` both true `endWritable` is not called.

I'm not sur what is wrong but what I found inside mailsplit is that the last not is the rfc822 and not the one with only the header (content-type: text/plain + subject: ok) and think it's the main issue.
I try to investigate mailsplit but I really don't know that library and I don't understand what is wrong, and why it stop on the parent node (rfc822 one).

I hop I will give you some insight. I will try again to search the origine of the issue but it's start to be hard to me to find time and to understand exactly what's wrong in mailsplit :/

**Comment by peernohell** [comment]
Hi,

I have find a fix on [https://github.com/andris9/mailsplit/pull/6](mailsplit) but as the maintener never give me a response I don't know if he will merge it :/

Also, if you want I have work on a monkey patch that can be added into simple-parser.
You just need to add this code after the last require:

```javascript
// monkey patch mailsplit.

const HEAD = 0x01;
const BODY = 0x02;
Splitter.prototype.processLine = function (line, final, next) {
    let flush = false;

    if (this.line && line) {
        line = Buffer.concat([this.line, line]);
        this.line = false;
    } else if (this.line && !line) {
        line = this.line;
        this.line = false;
    }

    if (!line) {
        line = Buffer.alloc(0);
    }

    if (this.nodeCounter > this.maxChildNodes) {
        let err = new Error('Max allowed child nodes exceeded');
        err.code = 'EMAXLEN';
        return next(err);
    }

    // we check boundary outside the HEAD/BODY scope as it may appear anywhere
    let boundary = this.checkBoundary(line);
    if (boundary) {
        // reached boundary, switch context
        switch (boundary) {
            case 1:
                // next child
                this.newNode(this.node);
                flush = true;
                break;
            case 2:
                // reached end of children, keep current node
                break;
            case 3: {
                // next sibling
                let parentNode = this.node.parentNode;
                if (parentNode && parentNode.contentType === 'message/rfc822') {
                    // special case where immediate parent is an inline message block
                    // move up another step
                    parentNode = parentNode.parentNode;
                }
                this.newNode(parentNode);
                flush = true;
                break;
            }
            case 4:
                // special case when boundary close a node with only header.
                if (this.node && this.node._headerlen && !this.node.headers) {
                    this.node.parseHeaders();
                    this.push(this.node);
                }
                // move up
                if (this.tree.length) {
                    this.node = this.tree.pop();
                }
                this.state = BODY;
                break;
        }

        return next(
            null,
            {
                node: this.node,
                type: 'data',
                value: line
            },
            flush
        );
    }

    switch (this.state) {
        case HEAD: {
            this.node.addHeaderChunk(line);
            if (this.node._headerlen > this.maxHeadSize) {
                let err = new Error('Max header size for a MIME node exceeded');
                err.code = 'EMAXLEN';
                return next(err);
            }
            if (final || (line.length === 1 && line[0] === 0x0a) || (line.length === 2 && line[0] === 0x0d && line[1] === 0x0a)) {
                let currentNode = this.node;

                currentNode.parseHeaders();

                // if the content is attached message then just continue
                if (
                    currentNode.contentType === 'message/rfc822' &&
                    !this.config.ignoreEmbedded &&
                    (!currentNode.encoding || ['7bit', '8bit', 'binary'].includes(currentNode.encoding)) &&
                    currentNode.disposition !== 'attachment'
                ) {
                    currentNode.messageNode = true;
                    this.newNode(currentNode);
                    if (currentNode.parentNode) {
                        this.node._parentBoundary = currentNode.parentNode._boundary;
                    }
                } else {
                    if (currentNode.contentType === 'message/rfc822') {
                        currentNode.messageNode = false;
                    }
                    this.state = BODY;
                    if (currentNode.multipart && currentNode._boundary) {
                        this.tree.push(currentNode);
                    }
                }

                return next(null, currentNode, flush);
            }

            return next();
        }
        case BODY: {
            return next(
                null,
                {
                    node: this.node,
                    type: this.node.multipart ? 'data' : 'body',
                    value: line
                },
                flush
            );
        }
    }

    next(null, false);
}
```

If you want me to push a pull request with the monkey patch I will be happy to do so.

**Comment by andris9** [comment]
Thanks! Published new version of mailparser with the fixed mailsplit dependency as v2.4.1
